API_DESIGN.md

```diff
@@ -0,0 +1,217 @@
+Check the possible status codes https://zitadel.com/docs/apis/statuscodes
+
+### Error Codes
+
+Additionally to the status code, the API returns unique error codes for each type of error.
+The error codes are used to identify a specific error and can be used to handle the error programmatically.
+
+> [!NOTE]
+> Currently, ZITADEL might already return some error codes. However, they do not follow a specific pattern yet
+> and are not documented. We will update the error codes and document them in the future.
+
+### Error Message and Details
+
+The API returns additional details about the error in the response body.
+This includes a human-readable error message and additional information that can help the client to understand the error
+as well as machine-readable details that can be used to handle the error programmatically.
+Error details use the Google RPC error details format: https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
+
+### Example
+
+HTTP/1.1 example:
+```
+HTTP/1.1 400 Bad Request
+Content-Type: application/json
+
+{
+  "code": "USER-001",
+  "message": "user requires a username",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.BadRequest",
+      "fieldViolations": [
+        {
+          "field": "username",
+          "description": "Username is required"
+        }
+      ]
+    }
+  ]
+}
+```
+
+gRPC / connectRPC example:
+```
+HTTP/2.0 200 Bad Request
+Content-Type: application/grpc
+Grpc-Message: user requires a username
+Grpc-Status: 3
+
+{
+  "code": "USER-001",
+  "message": "user requires a username",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.BadRequest",
+      "fieldViolations": [
+        {
+          "field": "username",
+          "description": "Username is required"
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Documentation
+
+- Document the purpose of the API, the services, the endpoints, the request and response messages, the error codes and the status codes.
+- Describe the fields of the request and response messages, the purpose and if needed the constraints.  
+- Document if the endpoints requires specific permissions or roles.
+- Document and explain the possible error codes and the error messages that can be returned by the API.
+
+#### Examples
+
+```proto
+// ListUsers will return all matching users. By default, we will return all users of your instance that you have permission to read. Make sure to include a limit and sorting for pagination.
+//
+// Required permission:
+//   - user.read
+//   - no permission required to own user
+//
+// Error Codes:
+//   - invalid_request: Your request does not have a valid format. Check error details for the reason.
+//   - unauthenticated: You are not authenticated. Please provide a valid access token.
+//   - permission_denied: You do not have the required permissions to access the requested resource.
```

Comment:
I think a table would make more sense here containing the following headers:

message, grpc code, http code, description

Reply:
updated as discussed